snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field
+from threading import Event, Lock
+from typing import Callable, Generic, Mapping, MutableMapping, Optional, Set, Sequence
+
+from snuba.utils.concurrent import execute
+from snuba.utils.streams.consumers.backends.abstract import ConsumerBackend
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
```

Comment:
Example usage:

```python
from sentry.utils.streams.consumers.backends import kafka
>>> kafka.TopicPartition('example', 0) in kafka.Topic('example')
True
>>> kafka.TopicPartition('bad-example', 0) in kafka.Topic('example')
False
>>> kafka.TopicPartition('^example-*', 0) in kafka.Topic('example-a')
True
>>> kafka.TopicPartition('^bad-example-*', 0) in kafka.Topic('example-a')
False
```

The regular expression syntax is defined by [`Consumer.subscribe`](https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.Consumer.subscribe).

In non-Kafka contexts, a "topic" could be extrapolated to be files in a directory for testing (the files themselves would be streams), etc.

In practical use, it might make sense to make these the same data structures that are used for `bootstrap` and the dataset configuration methods.